tests/deadcode.py

```diff
@@ -0,0 +1,269 @@
+            for stmt in node.body:
+                if isinstance(stmt, ast.ImportFrom):
+                    if stmt.module == "__future__" and any(
+                        "annotations" == alias.name for alias in stmt.names
+                    ):
+                        self.future_annotations = True
+
+            self.check_stmts(node.body, deadcode)
+        elif isinstance(node, (ast.With, ast.AsyncWith)):
+            self.check_childs(node.items, deadcode)
+            deadcode = self.check_stmts(node.body, deadcode)
+
+        elif isinstance(node, (ast.Return, ast.Break, ast.Continue, ast.Raise)):
+            if isinstance(node, ast.Raise):
+                self.walk_deadcode(node.exc, deadcode)
+                self.walk_deadcode(node.cause, deadcode)
+
+            if isinstance(node, ast.Return):
+                self.walk_deadcode(node.value, deadcode)
+
+            deadcode = True
+
+        elif isinstance(node, ast.Assert):
+            cnd = self.static_value(node.test, deadcode)
+
+            if cnd == False:
+                deadcode = True
+                self.walk_deadcode(node.msg, deadcode)
+
+            elif cnd == True:
+                node.deadcode = True
+                self.walk_deadcode(node.msg, True)
+                self.walk_deadcode(node.test, True)
+
+        elif isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
+            self.walk_annotation(node.args.posonlyargs, deadcode)
+            self.walk_annotation(node.args.args, deadcode)
+            self.walk_annotation(node.args.kwonlyargs, deadcode)
+
+            self.walk_annotation(node.args.vararg, deadcode)
+            self.walk_annotation(node.args.kwarg, deadcode)
+
+            self.check_childs(node.args.kw_defaults, deadcode)
+            self.check_childs(node.args.defaults, deadcode)
+
+            self.walk_annotation(node.returns, deadcode)
+            self.check_childs(node.decorator_list, deadcode)
+
+            self.check_stmts(node.body, deadcode)
+
+        elif isinstance(node, ast.ClassDef):
+            self.check_childs(node.decorator_list, deadcode)
+
+            self.check_childs(node.bases, deadcode)
+
+            self.check_childs(node.keywords, deadcode)
+
+            self.check_stmts(node.body, deadcode)
+
+        elif isinstance(node, ast.If):
+
+            test_value = self.static_value(node.test, deadcode)
+
+            if_is_dead = self.check_stmts(node.body, deadcode or (test_value == False))
+            else_is_dead = self.check_stmts(
+                node.orelse, deadcode or (test_value == True)
+            )
+            deadcode = if_is_dead and else_is_dead
+
+            self.walk_deadcode(node.test, (if_is_dead and not node.orelse) or deadcode)
+
+        elif isinstance(node, ast.Match):
+            self.walk_deadcode(node.subject, deadcode)
+            for _case in node.cases:
+                _case.deadcode = deadcode
+                self.walk_deadcode(_case.pattern, deadcode)
```

Comment:
I haven't really checked, but my vague understanding of pattern matching is that all patterns are essentially 'dead', i.e. they're syntax describing how to match/unpack/destructure rather than expressions that get evaluated. Which would explain why you can't find matching nodes.

Reply:
I think this will be the next bug report ... We and every python user have no way to know which attribute lookup caused the error.

script:
```python
class A:
    def __init__(self):
        pass

    def __getattr__(self, name):
        assert name == "foo", name


match A():
    case A(foo=5, bar=3):
        print("hi")
```

output (Python 3.11.0rc2+):
```python
Traceback (most recent call last):
  File "/home/frank/projects/executing/match_example.py", line 10, in <module>
    case A(foo=5, bar=3):
         ^^^^^^^^^^^^^^^
  File "/home/frank/projects/executing/match_example.py", line 6, in __getattr__
    assert name == "foo", name
AssertionError: bar
```